src/main/cpp/blaze_util_linux.cc

```diff
@@ -86,7 +87,7 @@ string GetSelfPath(const char* argv0) {
   // The file to which this symlink points could change contents or go missing
   // concurrent with execution of the Bazel client, so we don't eagerly resolve
   // it.
-  return "/proc/self/exe";
+  return std::string((char *)getauxval(AT_EXECFN));
```

Comment:
supernit: getauxval can technically return nullptr (maybe check that). Also consider a more C++ `reinterpret_cast<const char*>`.

Sorry for the silly question, but you mentioned running on M1 Mac in the description -- is the change in `blaze_util_linux.cc` sufficient for that? Or is that the case of docker running a Linux image on M1?

Reply:
I think the main question here is about potentially relative vs absolute paths (also why `/proc/self/exe` is not Bazel in this case). This is the path used to start the process so it can be the absolute path of the process but it also could be a symlink/relative path. The danger here is that if we end up changing the working directory at any point, `GetSelfPath` may not point us effectively to usable self anymore. Granted, I don't think we do that nor that we need that path a lot of times, but it is something to keep in mind.

Reply:
> Sorry for the silly question, but you mentioned running on M1 Mac in the description -- is the change in blaze_util_linux.cc sufficient for that? Or is that the case of docker running a Linux image on M1?

This is running a Linux/x86_64 container on Linux/arm. This is how Docker for Mac works under the hood. It runs a Linux VM running arm Linux and then installs runs the container in that with a binfmt_misc handler set up to run x86_64 binaries through qemu-user-static. This is useful because you usually want to run the same container and environment on your dev host as you do in your CI environment, rather than rebuilding everything for an extra platform. Technically, this isn't limited to running x86 on arm, it would also allow you to run an x86_64 bazel on say, MIPS, or an arm bazel on x86_64 using binfmt_misc. 

The underlying issue is that when you're using binfmt_misc, `/proc/self/exe` is qemu-user-static and not the bazel binary. getauxval(3) is a different api that provides similar information, but it's user-space memory that qemu-user-static can properly set up prior to handing off control to the binary (bazel in this case). arg[0] is another API that can be used which qemu-user-static will patch, but it's liable to contain links or path traversal such that resolving it is more work for us (`../../bazel`, etc).